packages/pipeline-editor/src/PipelineEditorWidget.tsx

```diff
@@ -681,16 +834,98 @@ export class PipelineEditor extends React.Component<
               }
             });
           }
-        } else {
-          // in this case, pipeline version is current
-          this.canvasController.setPipelineFlow(pipelineJson);
         }
       }
       this.setState({ emptyPipeline: Utils.isEmptyPipeline(pipelineJson) });
+      this.canvasController.setPipelineFlow(pipelineJson);
+      this.invalidPipeline();
     });
   }
 
+  // Adds an error decoration if a node has any invalid properties.
+  // Returns true if the node is valid.
+  updateDecorations(node: any): boolean {
+    node.app_data.invalidNodeError = this.invalidProperties(node);
+    if (node.app_data.invalidNodeError != null) {
+      this.canvasController.setNodeDecorations(node.id, [
+        {
+          id: 'error',
+          image: IconUtil.encode(errorIcon),
+          outline: false,
+          class_name: 'elyra-canvasErrorIcon',
+          position: 'topLeft',
+          x_pos: 20,
+          y_pos: 3
+        }
+      ]);
+      const pipelineId = this.canvasController.getPrimaryPipelineId();
+      const stylePipelineObj: any = {};
+      stylePipelineObj[pipelineId] = [node.id];
+      const styleSpec = {
+        body: { default: 'stroke: #d32f2f;' },
+        label: { default: 'fill: #d32f2f;' }
+      };
+      this.canvasController.setObjectsStyle(stylePipelineObj, styleSpec, true);
+      return false;
+    } else {
+      this.canvasController.setNodeDecorations(node.id, []);
+      return true;
+    }
+  }
+
+  /**
+   * Validates the properties of a given node.
+   *
+   * @param node: node to check properties for
+   *
+   * @returns a warning message to display in the tooltip
+   * if there are invalid properties. If there are none,
+   * returns null.
+   */
+  invalidProperties(node: any): string {
+    if (
+      node.app_data.runtime_image == null ||
+      node.app_data.runtime_image == ''
+    ) {
+      return 'no runtime image.';
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Validates the properties of all nodes in the pipeline.
+   * Updates the decorations / style of all nodes.
+   *
+   * @returns true if the pipeline is invalid.
+   */
+  invalidPipeline(): boolean {
```

Comment:
the name of this function is not accurate to what it does. From the name and return value I would assume this returns if a pipeline is invalid, but the actual function does more than that, updating the pipeline base on the state's validity. This function should either be renamed to make its actual usage clear, or split into two separate functions, one for the update, and one for the boolean return.

Reply:
Since I do the same thing in `updateDecorations` (i.e. both update the decorations and return the validity of the node) do you think maybe it should be called `updateAllDecorations`? And it just returns the validity of the pipeline?